crates/hir-ty/src/method_resolution.rs

```diff
@@ -646,30 +651,31 @@ pub fn lookup_impl_method(
 fn find_matching_impl(
     mut impls: impl Iterator<Item = ImplId>,
     table: &mut InferenceTable<'_>,
-    self_ty: &Ty,
+    expected_subst: Substitution,
 ) -> Option<Arc<ImplData>> {
     let db = table.db;
     loop {
         let impl_ = impls.next()?;
         let r = table.run_in_snapshot(|table| {
             let impl_data = db.impl_data(impl_);
-            let substs =
+            let impl_substs =
                 TyBuilder::subst_for_def(db, impl_, None).fill_with_inference_vars(table).build();
-            let impl_ty = db.impl_self_ty(impl_).substitute(Interner, &substs);
+            let trait_ref = db
+                .impl_trait(impl_)
+                .expect("non-trait method in find_matching_impl")
+                .substitute(Interner, &impl_substs);
 
-            table
-                .unify(self_ty, &impl_ty)
-                .then(|| {
-                    let wh_goals =
-                        crate::chalk_db::convert_where_clauses(db, impl_.into(), &substs)
-                            .into_iter()
-                            .map(|b| b.cast(Interner));
-
-                    let goal = crate::Goal::all(Interner, wh_goals);
+            if !table
+                .unify(trait_ref.substitution.as_slice(Interner), expected_subst.as_slice(Interner))
```

Comment:
maybe we should just have actual `TraitRef`s here, then we could just unify those